ygen/protogen.go

```diff
@@ -639,9 +649,23 @@ func safeProtoIdentifierName(name string) string {
 // between 1 and 2^29-1. The values 19,000-19,999 are excluded as these are explicitly
 // reserved for protobuf-internal use by https://developers.google.com/protocol-buffers/docs/proto3.
 func protoTagForEntry(e *yang.Entry) (uint32, error) {
-	// TODO(robjs): Replace this function with the final implementation
-	// once concluded.
-	return 1, nil
+	return fieldTag(e.Path())
+}
+
+// fieldTag takes an input string and calculates a FNV hash for the value. If the
+// hash is in the range 19,000-19,999 or 1-1,000, the input string has _ appended to
+// it and the hash is calculated.
+func fieldTag(s string) (uint32, error) {
+	h := fnv.New32()
+	if _, err := h.Write([]byte(s)); err != nil {
+		return 0, fmt.Errorf("could not write field path to hash: %v", err)
+	}
+
+	v := h.Sum32() & 0x1fffffff
```

Comment:
add // 2^29-1?
To be squeaky clean I recommend defining it as a package const maxProtoTagValue = 1 << 29 -1 and putting the blurb about range there.

Reply:
It seems like this splits the definition of the allowed ranges across multiple places (shouldn't we then make 19,000-19,999 a const somewhere too? My view here has been only use const where it might be referenced in >1 place, or it's a clear design choice in the package.